### PR TITLE
Add publics method to pipeline

### PR DESCRIPTION
**Pull request by Schaeff**
This seemed to me like it was missing.

### CODE REVIEW

File: pipeline/src/pipeline.rs [hunk 1]
```diff
@@ -827,6 +828,12 @@ impl<T: FieldElement> Pipeline<T> {
         Ok(self.artifact.witness.as_ref().unwrap().clone())
     }
 
+    pub fn publics(&self) -> Result<Vec<(String, T)>, Vec<String>> {
+        let pil = self.optimized_pil()?;
```
Comment (Schaeff) [inline pipeline/src/pipeline.rs:832]: Is `optimized_pil` the right method to call here? And should this method take a mutable reference and compute anything that isn't there yet?

Comment (leonardoalt) [reply pipeline/src/pipeline.rs:832]: The `compute_*` methods take a `&mut self` and actually compute whatever is needed. `optimized_pil` takes an immutable ref an assumes all the needed deps have been computed.

But don't you need the witness here too?

File: pipeline/src/pipeline.rs [hunk 2]
```diff
@@ -827,6 +828,12 @@ impl<T: FieldElement> Pipeline<T> {
...
+        let witness = self.witness()?;
```
Comment (leonardoalt) [inline pipeline/src/pipeline.rs:833]: witness has optimized_pil as dependency, so if you assume the witness exists then optimized_pil should exist too

File: pipeline/src/pipeline.rs (hunk 1 above)
Comment (Schaeff) [reply pipeline/src/pipeline.rs:832]: Yes I get it on the next line

File: pipeline/src/pipeline.rs [hunk 3]
```diff
@@ -827,6 +828,12 @@ impl<T: FieldElement> Pipeline<T> {
         Ok(self.artifact.witness.as_ref().unwrap().clone())
     }
 
+    pub fn publics(&self) -> Result<Vec<(String, T)>, Vec<String>> {
```
Comment (chriseth) [inline pipeline/src/pipeline.rs:831]: Wait but what does this do? Aren't publics always inputs?

Comment (leonardoalt) [reply pipeline/src/pipeline.rs:831]: This retrieves the publics from the witness. It's useful for backends that need to pass the publics explicitly for the prover or a verifier

Comment (leonardoalt) [reply pipeline/src/pipeline.rs:831]: The user passes the publics to powdr/pipeline, but for powdr they're just part of the witness table, with some copy constraints.

Comment (chriseth) [reply pipeline/src/pipeline.rs:831]: Maybe this should be documented? I would not have expected a full witness computation just from asking for the publics.

Comment (leonardoalt) [reply pipeline/src/pipeline.rs:831]: But thats not it, its just for when you already have the witness but just want to know what the publics are

**Review by leonardoalt — COMMENTED** [review]
I think the current assumptions here are fine

**Review by leonardoalt — APPROVED** [review]

**Review by leonardoalt — APPROVED** [review]

**Pull request merged**